### PR TITLE
Add completion status

### PR DESCRIPTION
**Pull request by cwiklik**

### CODE REVIEW

File: pkg/controller/queuejob/queuejob_controller_ex.go [hunk 1]
```diff
@@ -1012,17 +1047,17 @@ func (qjm *XController) getAggregatedAvailableResourcesPriority(unallocatedClust
...
-        apath := path.Dir(qjm.agentList[0]) 
+		apath := path.Dir(qjm.agentList[0])
 		var agentIdList = make([]string, len(clusters))
-		clustersProvided := false  // assume clusters not provided
+		clustersProvided := false // assume clusters not provided
 		for _, clusterRef := range clusters {
 			if clusterRef.Name != "" {
 				clustersProvided = true
-				agentIdList = append(agentIdList, apath+"/"+clusterRef.Name )
+				agentIdList = append(agentIdList, apath+"/"+clusterRef.Name)
```
Comment (dmatch01) [inline pkg/controller/queuejob/queuejob_controller_ex.go:1060]: Will this logic work with the original implementation of the cluster list selection?  The original implementation that does not use KubeStellar.

File: pkg/controller/queuejob/queuejob_controller_ex.go [hunk 2]
```diff
@@ -2123,12 +2159,14 @@ func (cc *XController) manageQueueJob(qj *arbv1.AppWrapper, podPhaseChanges bool
...
 		} else if qj.Status.CanRun && qj.Status.State == arbv1.AppWrapperStateActive {
 			//set appwrapper status to Complete or RunningHoldCompletion
-			derivedAwStatus := cc.getAppWrapperCompletionStatus(qj)
+			derivedAwStatus, genericItemsCompletionStatus := cc.getAppWrapperCompletionStatus(qj)
 
 			//Set Appwrapper state to complete if all items in Appwrapper
 			//are completed
 			if derivedAwStatus == arbv1.AppWrapperStateRunningHoldCompletion {
 				qj.Status.State = derivedAwStatus
+				klog.V(1).Infof("[>>>>>>>>>>] Setting ItemCompletionStatus 1")
```
Comment (dmatch01) [inline pkg/controller/queuejob/queuejob_controller_ex.go]: nit: Likely don't need the extra formatting here `>>>>>>`.   Suggest the function name is provided instead.

File: pkg/controller/queuejob/queuejob_controller_ex.go [hunk 3]
```diff
@@ -2144,8 +2182,11 @@ func (cc *XController) manageQueueJob(qj *arbv1.AppWrapper, podPhaseChanges bool
 				}
 				cc.updateEtcd(updateQj, "[syncQueueJob] setRunningHoldCompletion")
 			}
+			
 			//Set appwrapper status to complete
 			if derivedAwStatus == arbv1.AppWrapperStateCompleted {
+				klog.V(1).Infof("[>>>>>>>>>>] Setting ItemCompletionStatus 2")
```
Comment (dmatch01) [inline pkg/controller/queuejob/queuejob_controller_ex.go]: Same comment as above.

File: pkg/controller/queuejob/queuejob_controller_ex.go [hunk 4]
```diff
@@ -2165,35 +2206,23 @@ func (cc *XController) manageQueueJob(qj *arbv1.AppWrapper, podPhaseChanges bool
...
+            }
 			// Bugfix to eliminate performance problem of overloading the event queue.
 		} else if podPhaseChanges { // Continued bug fix
 			// Only update etcd if AW status has changed.  This can happen for periodic
 			// updates of pod phase counts done in caller of this function.
 			if err := cc.updateEtcd(qj, "manageQueueJob - podPhaseChanges"); err != nil {
 				klog.Errorf("[manageQueueJob] Error updating etc for AW job=%s Status=%+v err=%+v", qj.Name, qj.Status, err)
 			}
+			// Finish adding qj to Etcd for dispatch
+		//	return nil
```
Comment (dmatch01) [inline pkg/controller/queuejob/queuejob_controller_ex.go]: Likely need to leave `return nil` to not generate a side effect where the return is `return err`.

**Review by dmatch01 — CHANGES_REQUESTED** [review]
This is looking great.  A couple more comments and please add a e2e test and run full e2e test locally/test server.

File: pkg/controller/queuejob/queuejob_controller_ex.go (hunk 1 above)
Comment (cwiklik) [reply pkg/controller/queuejob/queuejob_controller_ex.go:1060]: Yes, this code is only active when running with externalDispatch=true:

	if qjm.serverOption.ExternalDispatch { ... return agentId }

Right beyond this block is the original code which I did not change

File: pkg/controller/queuejob/queuejob_controller_ex.go (hunk 4 above)
Comment (cwiklik) [reply pkg/controller/queuejob/queuejob_controller_ex.go]: The return nil is commented out in the original code (1.31). Not something I've changed.

**Pull request merged**